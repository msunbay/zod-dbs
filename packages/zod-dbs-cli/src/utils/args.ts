export const getArgumentValue = (name: string): string | undefined => {
  const argv = process.argv.slice(2);

  const key = name.startsWith('--') ? name : `--${name}`;
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === key) return argv[i + 1];
    if (arg.startsWith(`${key}=`)) return arg.slice(key.length + 1);
  }
  return undefined;
};

export const hasArgument = (name: string): boolean => {
  const argv = process.argv.slice(2);
  const key = name.startsWith('--') ? name : `--${name}`;
  return argv.includes(key) || argv.some((arg) => arg.startsWith(`${key}=`));
};

export const isSilentMode = (): boolean => {
  return hasArgument('--silent');
};
