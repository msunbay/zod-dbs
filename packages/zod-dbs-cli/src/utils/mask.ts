const maskConnectionString = (connectionString: string): string => {
  if (!connectionString) return '';

  let masked = connectionString;

  // 1) Try URL-style masking: scheme://username:password@host ...
  // Works for any scheme (postgres, mysql, mongodb+srv, mssql, snowflake, etc.)
  try {
    // Only attempt URL parsing if it looks like a URL (has a scheme:)
    if (/^[a-zA-Z][a-zA-Z0-9+.-]*:/.test(masked)) {
      // If it contains authority (//), URL will preserve userinfo
      if (masked.includes('//')) {
        const url = new URL(masked);
        if (url.password) {
          url.password = '****';
        }
        // Normalize URL (might reorder params), then continue with further masking
        masked = url.toString();
      }
    }
  } catch {
    // Ignore parse errors; fall back to regex approaches below
  }

  // 2) Generic userinfo masking without full URL parse: mask username:password@ (any scheme)
  // Replace only the password portion after the first ':' in the userinfo segment
  masked = masked.replace(
    /(\/\/)([^\/?#@]*)(@)/,
    (_m, p1: string, userinfo: string, p3: string) => {
      const idx = userinfo.indexOf(':');
      if (idx === -1) return `${p1}${userinfo}${p3}`;
      const userPart = userinfo.slice(0, idx + 1); // include colon
      return `${p1}${userPart}****${p3}`;
    }
  );

  // 3) Mask sensitive query parameters in URLs: ?password=...&token=...
  masked = masked.replace(
    /([?&])(password|pwd|token|sessiontoken|secret|clientsecret|client[_-]?secret|accesskeyid|secretaccesskey|access[_-]?key|secret[_-]?key|api[_-]?key|accountkey)=([^&#]*)/gi,
    (_m, sep: string, key: string) => `${sep}${key}=****`
  );

  // 4) Mask DSN-style key=value pairs (e.g., mssql/oracle). Apply only on DSN-like strings
  // to avoid clobbering URL query strings.
  if (
    masked.includes(';') ||
    (!/^[a-zA-Z][a-zA-Z0-9+.-]*:/.test(masked) && !masked.includes('?'))
  ) {
    masked = masked.replace(
      /(\b(password|pwd|token|sessiontoken|secret|clientsecret|client[_-]?secret|accesskeyid|secretaccesskey|access[_-]?key|secret[_-]?key|api[_-]?key|accountkey)\s*=\s*)('(?:[^']*)'|"(?:[^"]*)"|[^;&\s'"\)]+)/gi,
      (_m, prefix: string, _key: string, value: string) => {
        if (value?.startsWith("'") && value.endsWith("'"))
          return `${prefix}'****'`;
        if (value?.startsWith('"') && value.endsWith('"'))
          return `${prefix}"****"`;
        return `${prefix}****`;
      }
    );
  }

  return masked;
};

export const maskSensitiveValue = (
  name: string,
  value: string | undefined
): string => {
  if (!value) return '';

  const lowerName = name.toLowerCase();

  if (
    lowerName.includes('password') ||
    lowerName.includes('secret') ||
    lowerName.includes('token')
  ) {
    return '****';
  }

  if (name === 'connectionString') {
    return maskConnectionString(value);
  }

  return value;
};
