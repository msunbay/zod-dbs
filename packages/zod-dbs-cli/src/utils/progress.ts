import mustache from 'mustache';
import ora from 'ora';

const PROGRESS_STATUS: Record<string, string> = {
  connecting: 'Connecting to database...',
  fetchingSchema: 'Fetching schema information...',
  generating: 'Generating {{total}} Zod schemas...',
  done: 'Done',
};

const silentProgressHandler = {
  onProgress: () => {}, // No-op in silent mode
  done: () => {},
  fail: () => {},
};

export const createProgressHandler = (silent?: boolean) => {
  if (silent) return silentProgressHandler;

  const spinner = ora();

  return {
    onProgress: (status: string, args?: unknown) => {
      if (spinner.isSpinning) spinner.succeed();
      spinner.start(mustache.render(PROGRESS_STATUS[status] || status, args));
    },
    done: () => {
      spinner.succeed();
    },
    fail: () => {
      spinner.fail();
    },
  };
};
