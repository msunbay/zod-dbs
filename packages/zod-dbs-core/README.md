# zod-dbs-core

Shared core functionality for [zod-dbs](https://github.com/msunbay/zod-dbs).
