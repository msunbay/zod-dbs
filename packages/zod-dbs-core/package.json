{
  "name": "zod-dbs-core",
  "version": "1.0.13",
  "description": "Shared core functionality for zod-dbs packages",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "type": "module",
  "scripts": {
    "build": "rm -rf ./dist && tsc --project tsconfig.build.json",
    "test": "pnpm run test:types && pnpm run test:unit",
    "test:unit": "vitest --project unit run",
    "test:types": "tsc --noEmit"
  },
  "author": "msunbay",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/msunbay/zod-dbs.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist"
  ],
  "dependencies": {
    "debug": "^4.4.1"
  }
}
