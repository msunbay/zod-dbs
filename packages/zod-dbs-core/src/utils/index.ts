export * from './connectionString.js';
export * from './debug.js';
export * from './error.js';
export * from './sql.js';
