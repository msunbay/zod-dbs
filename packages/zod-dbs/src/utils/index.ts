export * from './casing.js';
export * from './fs.js';
export * from './mustache.js';
export * from './singularize.js';
