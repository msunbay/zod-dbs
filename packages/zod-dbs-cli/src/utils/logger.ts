import chalk from 'chalk';
import { ZodDbsProvider } from 'zod-dbs-core';

import { maskSensitiveValue } from './mask.js';

export const logSetting = (name: string, value: string | boolean | object) => {
  let displayValue = value.toString();

  if (name === 'provider' && typeof value === 'object' && value !== null) {
    const provider = value as ZodDbsProvider;
    const name = provider.name ?? 'custom provider';

    displayValue = provider.displayName
      ? `${name} (${provider.displayName})`
      : name;
  }

  console.info(
    `- ${chalk.white(name)}: ${chalk.blue(maskSensitiveValue(name, displayValue))}`
  );
};

export const logAppName = (message: string) => {
  console.info(chalk.magenta(`\n${message}\n`));
};

export const logError = (message: string) => {
  console.error(chalk.red(`${message}`));
};

export const logWarning = (message: string) => {
  console.warn(`⚠️ ${message}`);
};
