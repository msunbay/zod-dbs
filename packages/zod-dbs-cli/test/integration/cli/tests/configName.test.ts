import { execSync } from 'child_process';
import { toError } from 'zod-dbs-core';

import { getCliPath } from '../../utils.js';

const cliPath = getCliPath();

it('CLI accepts custom config file name', async () => {
  const output = execSync(`node ${cliPath} --config-name test`, {
    stdio: 'pipe',
  });

  expect(output.toString()).toMatchSnapshot();
});

it('CLI fails if custom config file name does not exist', async () => {
  let error: Error | undefined;

  try {
    execSync(`node ${cliPath} --config-name nonexisting`, {
      stdio: 'pipe',
    });
  } catch (err) {
    error = toError(err);
  }

  expect(error).toBeDefined();
  expect(error!.message).toContain(
    'Failed to load configuration from file "zod-dbs-nonexisting"'
  );
});
