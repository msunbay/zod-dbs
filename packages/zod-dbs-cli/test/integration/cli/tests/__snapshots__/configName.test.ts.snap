// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI accepts custom config file name 1`] = `
"
zod-dbs CLI v1.0.21

- provider: test-provider (Test Provider)
- outputDir: ./test/integration/output/customConfigFile
- include: users,posts
- exclude: ^temp_
- zodVersion: 4
- stringifyDates: true
- defaultEmptyArray: true
- configName: test

"
`;
